runtime/src/iree/io/stream.c

```diff
@@ -0,0 +1,609 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/io/stream.h"
+
+// This is arbitrary - we should allow dynamic block sizes and such but keeping
+// this small only requires a reasonable fixed size stack alloc.
+#define IREE_IO_STREAM_COPY_BLOCK_SIZE (16 * 1024)
+
+//===----------------------------------------------------------------------===//
+// Stream utilities
+//===----------------------------------------------------------------------===//
+
+static const char* iree_io_stream_seek_mode_string(
+    iree_io_stream_seek_mode_t seek_mode) {
+  switch (seek_mode) {
+    case IREE_IO_STREAM_SEEK_SET:
+      return "set";
+    case IREE_IO_STREAM_SEEK_FROM_CURRENT:
+      return "from-current";
+    case IREE_IO_STREAM_SEEK_FROM_END:
+      return "from-end";
+    default:
+      return "?";
+  }
+}
+
+static iree_string_view_t iree_io_stream_mode_format(
+    iree_io_stream_mode_t value, iree_bitfield_string_temp_t* out_temp) {
+  static const iree_bitfield_string_mapping_t mappings[] = {
+      // Separate:
+      {IREE_IO_STREAM_MODE_READABLE, IREE_SVL("READABLE")},
+      {IREE_IO_STREAM_MODE_WRITABLE, IREE_SVL("WRITABLE")},
+      {IREE_IO_STREAM_MODE_RESIZABLE, IREE_SVL("RESIZABLE")},
+      {IREE_IO_STREAM_MODE_SEEKABLE, IREE_SVL("SEEKABLE")},
+      {IREE_IO_STREAM_MODE_MAPPABLE, IREE_SVL("MAPPABLE")},
+  };
+  return iree_bitfield_format_inline(value, IREE_ARRAYSIZE(mappings), mappings,
+                                     out_temp);
+}
+
+// Validates that all bits in |required_mode| are available in |allowed_mode|.
+static iree_status_t iree_io_stream_validate_mode(
+    iree_io_stream_mode_t allowed_mode, iree_io_stream_mode_t required_mode) {
+  if (!iree_all_bits_set(allowed_mode, required_mode)) {
+#if IREE_STATUS_MODE
+    iree_bitfield_string_temp_t temp0, temp1;
+    iree_string_view_t required_mode_str =
+        iree_io_stream_mode_format(required_mode, &temp0);
+    iree_string_view_t allowed_mode_str =
+        iree_io_stream_mode_format(allowed_mode, &temp1);
+    return iree_make_status(
+        IREE_STATUS_PERMISSION_DENIED,
+        "operation requires mode %.*s but stream was opened with %.*s",
+        (int)required_mode_str.size, required_mode_str.data,
+        (int)allowed_mode_str.size, allowed_mode_str.data);
```

Comment:
Can any of these strings be empty? Want to put quotes around `%.*s` to make it clear when one is empty string?
```suggestion
        "operation requires mode '%.*s' but stream was opened with '%.*s'",
        (int)required_mode_str.size, required_mode_str.data,
        (int)allowed_mode_str.size, allowed_mode_str.data);
```

Reply:
empty strings are not possible, but added them anyway to highlight they are dynamic values